README.md
# node-sheets

> Read rows from google spreadsheet with google's sheets api.

[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

## Installation

```bash
$ npm install node-sheets --save
$ yarn add node-sheets
```

## Usage

Example to retrieve data from [this google spreadsheet](https://docs.google.com/spreadsheets/d/1amfst1WVcQDntGe6walYt-4O5SCrHBD5WntbjhvfIm4) using ES7 async/await.

```javascript
import Sheets from 'node-sheets';
try {
  const gs = new Sheets('1amfst1WVcQDntGe6walYt-4O5SCrHBD5WntbjhvfIm4');
  const authData = require('someGoogleCredentials.json'); // authData = { client_email, private_key }
  await gs.authorizeJWT(authData);
  const table = await gs.tables('Formats!A1:E3');
  console.log(table.headers);
  console.log(table.formats);
  console.log(table.rows);
} catch (err) {
  console.error(err);
}
```

You can also use the lib with Promises.

```javascript
import Sheets from 'node-sheets';
const gs = new Sheets('1amfst1WVcQDntGe6walYt-4O5SCrHBD5WntbjhvfIm4');
const authData = require('someGoogleCredentials.json'); // authData = { client_email, private_key }
gs.authorizeJWT(authData)
  .then(() => gs.tables('Formats!A1:E3'))
  .then((table) => {
    console.log(table.headers);
    console.log(table.formats);
    console.log(table.rows);
  })
  .catch((err) => {
    console.error(err);
  });
```

If you want to use this with `require` you need to import the `default`:

```javascript
const Sheets = require('node-sheets').default;
```

## API

### `Sheets.tables(string|object|array)`

Returns tabular sheet data for the specified ranges. This method accepts three distinct type of arguments: string, object and array.

### String

If a **string** argument is specified, it defines the name of the range ([A1 notation](https://developers.google.com/sheets/api/guides/concepts#a1_notation)) to be retrieved from the spreadsheet.
The return model is a `SheetTable` object.

### Object

If an **object** argument is specified, we expect to have an object with `name` and (optional) `range` properties.
The return model is a `SheetTable` object.

### Array

When the argument is an **array**, we want to retrieve table values for several sheets.
The return model is an array of `SheetTable` objects.

### `SheetTable` response schema

The `.tables()` method returns SheetTable objects that contains tabular data for a sheet.

| Header 1   | Header 2 | Header 3 |
| ---------- | -------- | -------- |
| row 1 text | $0.41    | 3.00     |
| ...        | ...      | ...      |

```javascript
const table = await gs.tables('Formats');

{
 title: 'Formats',                                                        // name of the sheet/table
 headers: ['Header 1', 'Header 2', 'Header 3'],                           // name of the headers (1st row)
 formats: [                                                               // array with information regarding cell format
   { numberFormat: { type: 'NONE' } },
   { numberFormat: { type: 'CURRENCY', pattern: '"$"#,##0.00' } },
   { numberFormat: { type: 'NUMBER', pattern: '#,##0.00' } } ]
 rows: [                                                                  // rows contains the values for 2nd row ahead
   {                                                                      // Each row object has:
     'Header 1': { value: 'row 1 text', stringValue: 'row 1 text' },
     'Header 2': { value: 0.41, stringValue: '$0.41' },
     'Header 3': { value: 3, stringValue: '3.00' }
    },
   { ... },
   { ... }
 ]
}
```

Sample access to the value of col 'Header 2' of first row:

```javascript
const currencyValue = table.rows[0]['Header 2'].value; // 0.41
```

**Note:** Formats are retrieved from first data row.

### Sample usage

```js
const sheet = await gs.tables('main'); // ranges = ['main']
const sheet = await gs.tables('A100'); // ranges = ['A100']  - that is the cell A100 and not the sheet A100
const sheet = await gs.tables({ sheet: 'main' }); // ranges = ['main!A:ZZZ']
const sheet = await gs.tables({ sheet: 'main', range: 'A1:B4' }); // ranges = ['main!A1:B4']
const sheets = await gs.tables([
  { sheet: 'main' },
  { sheet: 'D001', range: 'A1:D3' },
  { sheet: 'D002' },
]); // ranges = ['main!A:ZZZ', 'D001!A1:D3', 'D002!A:ZZZ']
```

#### Caveat

Parsing as a cell or a named range will take precedence over a sheet name when using **string** argument.
More info [here](http://stackoverflow.com/a/39641586).

## Authentication

node-sheets offers two authentication methods.

1. With JWT token (`.authorizeJWT(auth [, scopes])`) using `private_key` and `client_email`, and also allowing to set auth scopes. The default auth scope is https://www.googleapis.com/auth/spreadsheets.readonly.

1. With api key (`.authorizeApiKey(apikey)`) using an api key you have created in the [google developers console](https://console.developers.google.com).

## `Sheets.getLastUpdateDate()`

Returns a [ISO_8601](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString) compatible string with the last update date of the spreadsheet.
This can be used to check if a re-fetch is needed.

## `Sheets.getSheetsNames()`

Returns a list with all the names of the sheets in the spreadsheet.

## Examples

You can check the `/test/index.js` file for examples.

## License

This library is licensed under MIT. Full license text is available in [LICENSE](LICENSE).

## Contributing

We appreciate any contribution to this project. We keep a list of features and bugs in the [issue tracker](https://github.com/urbancups/node-sheets/issue).


CHANGELOG.md
# 1.2.0

- Merge PR #222 with updated dependencies
- Remove travis CI and adopt github actions

# 1.1.0

- Fix pull request #14 and issue #10

# 1.0.0

- First stable release


test/index.js
import assert from 'assert';
import util from 'util';
import Sheets from './../src';

const SPREADSHEET_TEST_ID = '1amfst1WVcQDntGe6walYt-4O5SCrHBD5WntbjhvfIm4';
const SPREADSHEET_JWT_KEY = require('./cred/node-sheets-test.json');
const SPREADSHEET_API_KEY = 'AIzaSyAwwdh_6ktghYF_AgP5pT9EfeiYWVCTr1Q';

describe('Sheets', function () {
  this.timeout(20000);

  describe('#constructor', function () {
    it('should exist and be a function', () => {
      assert.strictEqual(typeof Sheets, 'function');
    });

    it('should throw Error with empty spreadsheetId in the constructor', () => {
      assert.throws(() => {
        new Sheets();
      }, Error);
    });

    it('should accept a string with spreadsheetId', () => {
      assert.notStrictEqual(new Sheets(SPREADSHEET_TEST_ID), null);
    });
  });

  describe('#authorize', function () {
    const gs = new Sheets(SPREADSHEET_TEST_ID);

    it('should throw Error with no args in authorizeJWT', async () => {
      try {
        await gs.authorizeJWT();
        assert.fail();
      } catch (err) {
        assert.strictEqual(err.constructor, Error);
      }
    });

    it('should throw Error with no args in authorizeApiKey', async () => {
      try {
        await gs.authorizeJWT();
        assert.fail();
      } catch (err) {
        assert.strictEqual(err.constructor, Error);
      }
    });

    it('should authenticate with valid JWT credentials', async () => {
      await gs.authorizeJWT(SPREADSHEET_JWT_KEY);
    });

    it('should authenticate with valid API KEY credentials', async () => {
      await gs.authorizeApiKey(SPREADSHEET_API_KEY);
    });
  });

  describe('#getSheetsNames', () => {
    it('should return a list of sheets names', async () => {
      const gs = new Sheets(SPREADSHEET_TEST_ID);
      await gs.authorizeApiKey(SPREADSHEET_API_KEY);
      const names = await gs.getSheetsNames();
      assert.deepStrictEqual(names, [
        'Class Data',
        'Table with empty cells',
        'Formats',
        'Empty',
        'blue',
        'D001',
        'D002',
      ]);
    });
  });

  describe('#getLastUpdateDate', () => {
    it('should return a valid date string', async () => {
      // ex: 2016-09-21T15:37:15.250Z
      const gs = new Sheets(SPREADSHEET_TEST_ID);
      await gs.authorizeApiKey(SPREADSHEET_API_KEY);
      const updateDate = await gs.getLastUpdateDate();
      assert.notStrictEqual(updateDate, null);
      const date = new Date(updateDate);
      assert.strictEqual(date.constructor, Date);
    });
  });

  describe('#tables', () => {
    it('should retrieve all sheets', async () => {
      const gs = new Sheets(SPREADSHEET_TEST_ID);
      await gs.authorizeApiKey(SPREADSHEET_API_KEY);
      const names = await gs.getSheetsNames();
      const tables = await gs.tables(names.map((name) => ({ name: name })));
      //console.log(util.inspect(tables, { depth: null, colors: true }))
      assert.strictEqual(tables.length, names.length);
      assert.deepStrictEqual(
        tables.map((t) => t.title),
        names
      );

      // const table = tables[0]
      // console.log(util.inspect(table.headers, { depth: null, colors: true }))
      // console.log(util.inspect(table.formats, { depth: null, colors: true }))
      // console.log(util.inspect(table.rows, { depth: null, colors: true }))
    });
  });

  describe('#tables (Formats!A1:E3)', () => {
    const gs = new Sheets(SPREADSHEET_TEST_ID);
    before(async () => {
      await gs.authorizeApiKey(SPREADSHEET_API_KEY);
    });

    it('should return formatted tabular spreadsheet data', async () => {
      const table = await gs.tables('Formats!A1:E3');
      // console.log(util.inspect(table, { depth: null, colors: true }))
      assert.deepStrictEqual(table.headers, [
        'Automatic',
        'Currency',
        'Date',
        'Number',
        'Plain Text',
      ]);
      assert.deepStrictEqual(
        table.formats.map((f) => f.numberFormat.type),
        ['NONE', 'CURRENCY', 'DATE', 'NUMBER', 'TEXT']
      );
      assert.strictEqual(table.rows.length, 2);
      assert.deepStrictEqual(Object.keys(table.rows[0]), [
        'Automatic',
        'Currency',
        'Date',
        'Number',
        'Plain Text',
      ]);
      assert.deepStrictEqual(table.rows[0]['Automatic'], {
        value: 'Oil',
        stringValue: 'Oil',
      });
      assert.deepStrictEqual(table.rows[0]['Currency'], {
        value: 0.41,
        stringValue: '$0.41',
      });
      assert.deepStrictEqual(table.rows[0]['Date'], {
        value: new Date(2016, 0, 25),
        stringValue: '1/25/2016',
      });
      assert.deepStrictEqual(table.rows[0]['Number'], {
        value: 123,
        stringValue: '123.00',
      });
      assert.deepStrictEqual(table.rows[0]['Plain Text'], {
        value: 'This is some text',
        stringValue: 'This is some text',
      });
    });

    it('should return formatted tabular (tableCols) spreadsheet data', async () => {
      const cols = await gs.tableCols('Formats!A1:E3');
      assert.strictEqual(cols.length, 5);
      assert.deepStrictEqual(
        cols.map((c) => c.header),
        ['Automatic', 'Currency', 'Date', 'Number', 'Plain Text']
      );
      assert.strictEqual(cols[0].header, 'Automatic');
    });
  });

  describe('#table (Class Data)', () => {
    const gs = new Sheets(SPREADSHEET_TEST_ID);
    before(async () => {
      await gs.authorizeApiKey(SPREADSHEET_API_KEY);
    });

    it('should retrieve only cols and rows with content (8 cols and 2 rows - from "Class Data")', async () => {
      const table = await gs.tables('Class Data');
      assert.strictEqual(table.headers.length, 8);
      assert.strictEqual(table.rows.length, 2);
      // console.log(util.inspect(table, { depth: null, colors: true, breakLength: Infinity }))
    });

    it('should use named range instead of sheet name for special names', async () => {
      const table = await gs.tables('A02');
      // console.log(util.inspect(table, { depth: null, colors: true, breakLength: Infinity }))
      assert.strictEqual(table.rows.length, 0);
      assert.strictEqual(table.headers.length, 1);
      assert.strictEqual(table.headers[0], 'Thomas');
    });

    it('should retrieve empty cells as undefined (5 cols and 3 rows - from "Table with empty cells")', async () => {
      const table = await gs.tables('Table with empty cells');
      // console.log(util.inspect(table, { depth: null, colors: true, breakLength: Infinity }))
      assert.strictEqual(table.headers.length, 5);
      assert.strictEqual(table.rows.length, 3);
      assert.deepStrictEqual(
        Object.keys(table.rows[1]).map((col) => table.rows[1][col].value),
        ['Andrew', undefined, '1. Freshman', 'SD', 'Math']
      );
      assert.deepStrictEqual(
        Object.keys(table.rows[2]).map((col) => table.rows[2][col].value),
        ['Anna', 'Female', undefined, undefined, 'English']
      );
    });
  });

  describe('works with Promises', () => {
    it('should throw Error with no args in authorization', () => {
      const gs = new Sheets(SPREADSHEET_TEST_ID);
      gs.authorizeJWT()
        .then(() => assert.fail())
        .catch((err) => assert.strictEqual(err.constructor, Error));
    });

    it('should be able to chain .then() calls, and also .catch()', () => {
      const gs = new Sheets(SPREADSHEET_TEST_ID);
      const authData = SPREADSHEET_JWT_KEY;
      gs.authorizeJWT(authData)
        .then(() => gs.tables('Formats!A1:E3'))
        .then((table) => {
          assert.notStrictEqual(table.headers, null);
          assert.notStrictEqual(table.formats, null);
          assert.notStrictEqual(table.rows, null);
        })
        .catch((err) => {
          console.error(err);
        });
    });
  });
});


src/sheets.js
import Q from 'q';
import { google } from 'googleapis';
import util from 'util';
import zipObject from 'lodash.zipobject';

const SCOPE_DRIVE_READ = 'https://www.googleapis.com/auth/drive.readonly';
const SCOPE_SPREADSHEETS_READ =
  'https://www.googleapis.com/auth/spreadsheets.readonly';

export default class Sheets {
  /**
   * Create an instance of Sheets for a given google spreadsheet id
   */
  constructor(spreadsheetId) {
    if (spreadsheetId === undefined) {
      throw new Error('"spreadsheetId" is a required argument');
    }

    this.spreadsheetId = spreadsheetId;
    this.auth = null;
  }

  /**
   * Performs Google APIs authentication with the given JWT key info.
   * The key object need to have a `client_email` and `private_key` fields.
   *
   * You can create a key in the [google developers console](https://console.developers.google.com).
   */
  async authorizeJWT(
    key,
    scopes = [SCOPE_SPREADSHEETS_READ, SCOPE_DRIVE_READ]
  ) {
    try {
      if (!key) throw new Error('"key" is a required argument');

      this.auth = new google.auth.JWT(
        key.client_email,
        null,
        key.private_key,
        scopes,
        null
      );
      await Q.ninvoke(this.auth, 'authorize');
    } catch (err) {
      throw err;
    }
  }

  /**
   * Performs Google APIs authentication with the given API KEy.
   *
   * You can create a new key in the [google developers console](https://console.developers.google.com).
   */
  async authorizeApiKey(apikey) {
    if (!apikey) throw new Error('"apikey" is a required argument');
    this.auth = apikey;
  }

  /**
   * Returns the last update date of the spreadsheet.
   * Date is returned as a string in RFC 3339 date-time format. Example: 1985-04-12T23:20:50.52Z
   * This date-time format is parsable by Date constructor in Javascript and momentjs library.
   */
  async getLastUpdateDate() {
    var drive = google.drive('v3');
    const response = await Q.ninvoke(drive.files, 'get', {
      auth: this.auth,
      fileId: this.spreadsheetId,
      fields: 'modifiedTime',
    });
    const res = response.data;
    return res.modifiedTime;
  }

  /**
   * Returns a list with all the names of the sheets in the spreadsheet.
   */
  async getSheetsNames() {
    var sheets = google.sheets('v4');
    const response = await Q.ninvoke(sheets.spreadsheets, 'get', {
      auth: this.auth,
      spreadsheetId: this.spreadsheetId,
      fields: 'sheets/properties',
    });
    const res = response.data.sheets.map((sheet) => sheet.properties.title);
    return res;
  }

  /**
   * Returns a spreadsheet range in tabular cols format.
   * The tabular cols format returns the content by cols, and each col contains the values for each row.
   *
   * [ { header: 'Automatic',  values: [ 'Oil', 'Grass' ],                       stringValues: [ 'Oil', 'Grass' ],                      format: { numberFormat: { type: 'NONE' } } },
   *   { header: 'Currency',   values: [ 0.41, 1.3 ],                            stringValues: [ '$0.41', '$1.30' ],                    format: { numberFormat: { type: 'CURRENCY', pattern: '"$"#,##0.00' } } },
   *   { header: 'Date',       values: [ Date(24, 1, 2016), Date(15, 5, 2016) ], stringValues: [ '1/25/2016', '5/15/2016' ],            format: { numberFormat: { type: 'DATE', pattern: 'M/d/yyyy' } } },
   *   { header: 'Number',     values: [ 123, 4 ],                               stringValues: [ '123.00', '4.00' ],                    format: { numberFormat: { type: 'NUMBER', pattern: '#,##0.00' } } },
   *   { header: 'Plain Text', values: [ 'This is some text', 'Another text' ],  stringValues: [ 'This is some text', 'Another text' ], format: { numberFormat: { type: 'TEXT' } } }
   * ]
   *
   * Note: Formats are retrieved from first data row.
   */
  async tableCols(range) {
    const spreadsheet = await getRanges(this.auth, this.spreadsheetId, [range]);
    const sheet = spreadsheet.sheets[0]; // first (unique) sheet
    const gridData = sheet.data[0]; // first (unique) range
    const headers = gridData.rowData[0].values // first row (headers)
      .map((col) => formattedValue(col));

    const otherRows = gridData.rowData.slice(1);

    return headers.map((header, headerIdx) => ({
      header: header,
      stringValues: otherRows.map((row) =>
        formattedValue(row.values[headerIdx])
      ),
      values: otherRows.map((row) => effectiveValue(row.values[headerIdx])),
      format: effectiveFormat(otherRows[0].values[headerIdx]), // format of first data line/cell
    }));
  }

  /**
   * Returns spreadsheet ranges in tabular row format.
   * The tabular row format returns the content by rows, and each row contains the values for each column.
   *
   * | Header 1   | Header 2 | Header 3 |
   * | ---------- | -------- | -------- |
   * | row 1 text | $0.41    | 3.00     |
   * | ...        | ...      | ...      |
   *
   *
   * {
   *  title: 'Formats',                                                        // name of the sheet/table
   *  headers: ['Header 1', 'Header 2', 'Header 3'],                           // name of the headers (1st row)
   *  formats: [                                                               // array with information regarding cell format
   *    { numberFormat: { type: 'NONE' } },
   *    { numberFormat: { type: 'CURRENCY', pattern: '"$"#,##0.00' } },
   *    { numberFormat: { type: 'NUMBER', pattern: '#,##0.00' } } ]
   *  rows: [                                                                  // rows contains the values for 2nd row ahead
   *    {                                                                      // Each row object has:
   *        'Header 1': { value: 'row 1 text', stringValue: 'row 1 text' },
   *        'Header 2': { value: 0.41, stringValue: '$0.41' },
   *        'Header 3': { value: 3, stringValue: '3.00' }
   *     },
   *    { ... },
   *    { ... }
   *  ]
   * }
   *
   * Sample access to the value of col 'Header 2' of first row:
   * ```
   * const currencyValue = table.rows[0]['Header 2'].value     // 0.41
   * ```
   *
   * Note: Formats are retrieved from first data row.
   */
  async tables(ranges) {
    let single = false;
    if (typeof ranges === 'string') {
      // string
      ranges = [{ name: ranges }]; // NOTE: This version doesn't add range `A:ZZZ`
      single = true;
    } else if (Array.isArray(ranges) === false) {
      // object
      ranges = [{ name: ranges.name, range: ranges.range || 'A:ZZZ' }];
    } else {
      // Array
      ranges = ranges.map((range) => ({
        name: range.name,
        range: range.range || 'A:ZZZ',
      }));
    }

    // convert ranges to google-sheets ranges
    ranges = ranges.map((r) => `${r.name}${r.range ? `!${r.range}` : ''}`);

    const spreadsheets = await getRanges(this.auth, this.spreadsheetId, ranges);
    const res = spreadsheets.sheets.map(sheetToTable);
    return single ? res[0] : res;
  }
}

/**
 * Converter from google sheet format to node-sheets format
 */
function sheetToTable(sheet) {
  if (sheet.data.length === 0 || sheet.data[0].rowData === undefined) {
    return {
      title: sheet.properties.title,
      headers: [],
      formats: [],
      rows: [],
    };
  }
  const gridData = sheet.data[0]; // first (unique) range
  const headers = gridData.rowData[0].values // first row (headers)
    .map((col) => formattedValue(col));

  const otherRows = gridData.rowData.slice(1);

  const values =
    otherRows.length > 0
      ? otherRows[0].values
      : new Array(headers.length).fill({});

  return {
    title: sheet.properties.title,
    headers: headers,
    formats: values.map((value) => effectiveFormat(value)),
    rows: otherRows.map((row) =>
      zipObject(
        headers,
        (row.values || []).map((value) => ({
          value: effectiveValue(value),
          stringValue: formattedValue(value),
        }))
      )
    ),
  };
}

function formattedValue(value) {
  return value ? value.formattedValue : value;
}

function effectiveValue(value) {
  if (!value) return value;

  if (value.effectiveFormat === null || value.effectiveFormat === undefined) {
    return value.formattedValue;
  }

  if (value.effectiveFormat.numberFormat) {
    switch (value.effectiveFormat.numberFormat.type) {
      case 'TEXT':
        return value.effectiveValue && value.effectiveValue.stringValue
          ? value.effectiveValue.stringValue
          : '';
      case 'NUMBER':
        return value.effectiveValue && value.effectiveValue.numberValue
          ? value.effectiveValue.numberValue
          : 0;
      case 'CURRENCY':
        return value.effectiveValue && value.effectiveValue.numberValue
          ? value.effectiveValue.numberValue
          : 0;
      case 'DATE': // 'serial number' format
        return value.effectiveValue && value.effectiveValue.numberValue
          ? ExcelDateToJSDate(value.effectiveValue.numberValue)
          : new Date();
    }
  }

  return value.formattedValue;
}

function effectiveFormat(value) {
  if (value.effectiveFormat) {
    return value.effectiveFormat;
  }
  return { numberFormat: { type: 'NONE' } };
}

/**
 * Converting Excel Date Serial Number to Date using Javascript
 * Source: http://stackoverflow.com/a/16233621/336596
 * @param {[type]} serial a date value in "Excel Date Serial" format
 */
function ExcelDateToJSDate(serial) {
  var utc_days = Math.floor(serial - 25569);
  var utc_value = utc_days * 86400;
  var date_info = new Date(utc_value * 1000);

  var fractional_day = serial - Math.floor(serial) + 0.0000001;

  var total_seconds = Math.floor(86400 * fractional_day);

  var seconds = total_seconds % 60;

  total_seconds -= seconds;

  var hours = Math.floor(total_seconds / (60 * 60));
  var minutes = Math.floor(total_seconds / 60) % 60;

  return new Date(
    date_info.getFullYear(),
    date_info.getMonth(),
    date_info.getDate(),
    hours,
    minutes,
    seconds
  );
}

/**
 * Utility function to get a range from a spreadsheet
 */
async function getRanges(
  auth,
  spreadsheetId,
  ranges,
  fields = 'properties.title,sheets.properties,sheets.data(rowData.values.effectiveValue,rowData.values.formattedValue,rowData.values.effectiveFormat.numberFormat)'
) {
  var sheets = google.sheets('v4');
  const response = await Q.ninvoke(sheets.spreadsheets, 'get', {
    auth: auth,
    spreadsheetId: spreadsheetId,
    // https://developers.google.com/sheets/guides/concepts, https://developers.google.com/sheets/samples/sheet
    fields: fields,
    includeGridData: true,
    ranges: ranges,
  });
  const spreadsheet = response.data;
  return spreadsheet;
}
